i/p/requestrules,o/i/apparmorprompting: batch adding rule notices

olivercalder opened the pull request:
This PR is based on #16127

Instead of separately calling `notifyRule` for each rule when many rules require notices (such as on snapd startup), make `notifyRule` into `notifyRules` which takes a list of rules and records notices for all of them at once, only writing notices to disk once after all have been added.

This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-35903

olivercalder committed:
i/p/requestrules,o/i/apparmorprompting: batch adding rule notices

Instead of separately calling `notifyRule` for each rule when many rules
require notices (such as on snapd startup), make `notifyRule` into
`notifyRules` which takes a list of rules and records notices for all of
them at once, only writing notices to disk once after all have been
added.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

ZeyadYasser commented on interfaces/prompting/constraints.go:
```diff
@@ -355,11 +355,22 @@ func (c *RuleConstraints) MarshalJSON() ([]byte, error) {
+
+const (
+	NoPermsExpired PermExpirationStatus = iota
+	AnyPermsExpired
```
```suggestion
	SomePermsExpired
```

olivercalder commented on overlord/ifacestate/apparmorprompting/prompting.go:
```diff
@@ -110,16 +110,22 @@ func New(s *state.State) (m *InterfacesRequestsManager, retErr error) {
 	}
-	notifyRule := func(userID uint32, ruleID prompting.IDType, data map[string]string) error {
+	notifyRules := func(rules []*requestrules.Rule, data map[string]string) error {
 		// TODO: add some sort of queue so that notifyRule calls can return
```
This is intentional to ease merging this change with the PR which re-enables the prompting notice backend and deletes this comment.

ZeyadYasser reviewed (approved):
LGTM, Thank you!

pedronis reviewed (commented):
naming comment

olivercalder committed:
i/p/requestrules: batch expired rule notices when adding new rule

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
i/prompting: rename AnyPermsExpired to SomePermsExpired

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented on interfaces/prompting/requestrules/requestrules.go:
```diff
@@ -325,23 +323,15 @@ func (rdb *RuleDB) load() (retErr error) {
 	}
 
+	rdb.notifyRules(partiallyExpiredRules, nil)
 	expiredData := map[string]string{"removed": "expired"}
```
Good point. I've renamed to simply `data` as part of the new logic to batch expired rule notices when adding a new rule.

olivercalder reviewed (commented):
Addressed review comments

olivercalder commented:
This will conflict with https://github.com/canonical/snapd/pull/16194 once that lands. That PR has priority over this one. The prompting notice backend will need to be adjusted to add a method to record a slice of notice IDs with a given notice data.

olivercalder committed:
o/i/apparmorprompting: add multiple notices to prompting backend

Add a new method to the prompting notice backend which takes a slice of
notice information and then records notices in a batch, only releasing
the notice backend lock and saving the state to disk once after all
notices have been recorded.

If there is an error saving the state to disk, then all changes are
rolled back. If there is an error when recording one or more notices in
a batch, then any other notices in the batch continue to be recorded.
Such an error should only occur if there is an existing notice with the
same ID but a different user, which should not happen in the real world.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
o/i/apparmorprompting: fix bug in notice backend when re-recording notice

When re-recording a notice which was expired at the time, there was a
bug where the new notice would be added to the ID map, then all
previously-expired notice IDs were removed from the map, so the
newly-added ID would be removed as well. Fix that bug and add a test for
it.

Additionally, return early if there are no notices to be added, and
remove unnecessary debug log.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

natibek commented on overlord/ifacestate/apparmorprompting/noticebackend.go:
```diff
@@ -196,18 +197,101 @@ func (ntb *noticeTypeBackend) noticeKeyToID(key string) string {
+		} else {
+			rollbackStack = append(rollbackStack, rollback)
+		}
+	}
```
Should we add a check if all the calls to `ntb.doAddNotice` returned a non-nil error before attempting to save?
